server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.validate.builtin;
+
+import org.keycloak.validate.CompactValidator;
+import org.keycloak.validate.ValidationContext;
+import org.keycloak.validate.ValidationError;
+import org.keycloak.validate.ValidatorConfig;
+
+import java.util.regex.Pattern;
+
+public class EmailValidator implements CompactValidator {
+
+    public static final String ID = "email";
+
+    public static final EmailValidator INSTANCE = new EmailValidator();
+
+    public static final String MESSAGE_INVALID_EMAIL = "error-invalid-email";
+
+    // Actually allow same emails like angular. See ValidationTest.testEmailValidation()
+    private static final Pattern EMAIL_PATTERN = Pattern.compile(
+            "[a-zA-Z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*");
+
+    private EmailValidator() {
+    }
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
+            context.addError(new ValidationError(ID, inputHint, ValidationError.MESSAGE_INVALID_VALUE, input));
+            return context;
+        }
+
+        if (!EMAIL_PATTERN.matcher((String) input).matches()) {
```

Comment:
This is format validation, so value should be validated only if it is not null to be able to operate on optional fields.